java/src/jmri/jmrit/display/layoutEditor/LayoutTrack.java

```diff
@@ -181,40 +175,105 @@ protected boolean isConnectionType(int connectionType) {
-        log.error("virtual method: override in sub-classes and don't call super...].");
-        log.error("Invalid connection type " + connectionType); //I18IN
-        return center;
-    }
+    public abstract Point2D getCoordsForConnectionType(int connectionType);
 
-    public void reCheckBlockBoundary() {
-        log.error("virtual method: override in sub-classes and don't call super...].");
-    }
+    /**
+     * abstract method… subclasses should implement _IF_ they need to recheck their block boundaries
+     */
+    public abstract void reCheckBlockBoundary();
 
     /**
      * @return the bounds of this track
+     * (abstract: should be overridden by ALL subclasses)
+     */
+    public abstract Rectangle2D getBounds();
+
+    /**
+     * get the object connected to this track for the specified connection type
+     * @param connectionType the specified connection type
+     * @return the object connected to this slip for the specified connection type
+     * @throws jmri.JmriException
      */
-    public Rectangle2D getBounds() {
-        log.error("virtual method: override in sub-classes and don't call super...].");
-        return new Rectangle2D.Double();
+    public Object getConnection(int connectionType) throws jmri.JmriException {
+        Object result = null;
+        switch (connectionType) {
+            case POS_POINT: {
+                result = ((PositionablePoint)this).getConnection(connectionType);
+                break;
+            }
+            case TURNOUT_A:
+            case TURNOUT_B:
+            case TURNOUT_C:
+            case TURNOUT_D: {
+                result = ((LayoutTurnout)this).getConnection(connectionType);
+                break;
+            }
+            case LEVEL_XING_A:
+            case LEVEL_XING_B:
+            case LEVEL_XING_C:
+            case LEVEL_XING_D: {
+                result = ((LevelXing)this).getConnection(connectionType);
+                break;
+            }
+
+            case TRACK: {
+                result = ((TrackSegment)this).getConnection(connectionType);
+                break;
+            }
+
+            case SLIP_A:
+            case SLIP_B:
+            case SLIP_C:
+            case SLIP_D: {
+                result = ((LayoutSlip)this).getConnection(connectionType);
+                break;
+            }
+            default: {
+                if (connectionType >= TURNTABLE_RAY_OFFSET) {
+                    result = ((LayoutTurntable)this).getConnection(connectionType);
+                } else {
+                    log.error("Invalid connection type " + connectionType); //I18IN
+                    throw new jmri.JmriException("Invalid Point");
+                }
+                break;
+            }
+        }
+        return result;
+    }
+
+    /**
+     * @param connectionType
+     * @return true if the connection for this connection type is free
+     */
+    public boolean isDisconnected(int connectionType) {
```

Comment:
Mostly ok with that… except I only do one return statement per method… at the bottom. Always use braces and always "break" (and comment verbosely when I don't) (As per  [Stephen Maguire's](https://www.amazon.com/Steve-Maguire/e/B00DP4BEB2/ref=dp_byline_cont_pop_book_1) "Writing Solid Code". (Yeah, I'm that "old school". ;-)) Fixes for the above ASAP.